### ISSUE TITLE
[MNG-4189] Maven not picking up specific timestamped version dependency when a later timestamped version was downloaded and already present in the local repository

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#5944]
**[Maria Odea B. Ching](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oching)** opened **[MNG-4189](https://issues.apache.org/jira/browse/MNG-4189?redirect=false)** and commented

To reproduce this issue:
1. Create a project (let's call this projectA)  with a class named ClassA having a method named methodA(). Set the version as 1.0-SNAPSHOT and set uniqueVersion=true.
2. Deploy this in a remote repository
3. Create another project (let's call this projectB) which has a dependency on projectA. Set the dependency's version to the specific timestamped version when projectA was deployed in step 2. Create a class named ClassB and add a method which invokes ClassA's methodA().
4. Add your remote repository either in the settings or in the pom.
5. Build projectB. You will get a successful build.
6. Now go back to projectA and remove methodA() from classA.
7. Deploy projectA to the remote repository again.
8. Update the dependency version of projectA in projectB's pom.xml. Set it to the latest timestamp version.
9. Build projectB. Your build will fail because methodA() was removed.
10. Revert the dependency version of projectA in projectB's pom.xml. Set it to the same value you've set in step 3.
11. Build projectB. Your build will still fail even though you've set the correct version.

---

**Affects:** 2.0.9, 2.1.0

**Attachments:**
- [MNG-4189.patch](https://issues.apache.org/jira/secure/attachment/12712936/MNG-4189.patch) (_1.82 kB_)
- [mng-4189.zip](https://issues.apache.org/jira/secure/attachment/12712933/mng-4189.zip) (_14.98 kB_)
- [MNG-4189-core-integration-testing.patch](https://issues.apache.org/jira/secure/attachment/12712935/MNG-4189-core-integration-testing.patch) (_20.82 kB_)
- [MNG-4189-maven-2.1.x.patch](https://issues.apache.org/jira/secure/attachment/12712912/MNG-4189-maven-2.1.x.patch) (_6.79 kB_)

**Issue Links:**
- [MNG-5129](https://issues.apache.org/jira/browse/MNG-5129) Maven struggles while resolving locked snapshots by two or more simultaneously used projects.

- [MENFORCER-298](https://issues.apache.org/jira/browse/MENFORCER-298) bannedDependencies can break compilation against a timestamped snapshot

**Comment by jira-importer** [comment]
**[Maria Odea B. Ching](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oching)** commented

I'm still creating an IT for this, will attach it to this jira issue once completed..

**Comment by jira-importer** [comment]
**[Maria Odea B. Ching](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oching)** commented

Attaching IT for this bug..

**Comment by jira-importer** [comment]
**[Maria Odea B. Ching](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oching)** commented

Attaching patch for core-integration-testing integrating the IT test for this issue..

**Comment by jira-importer** [comment]
**[Maria Odea B. Ching](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oching)** commented

Attaching MNG-4189-maven-2.1.x.patch to fix this issue. Unit tests included in the patch. Could someone please review this before I commit it to 2.1.x branch? Thanks in advance!

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

With your patch, we effectively have

```java
boolean resolved = false;
if ( !destination.exists() || force )
{
    ...
    resolved = true;
}
else if ( destination.exists() )     // Note that this is equivalent to just "else"
{
    ...
    resolved = true;
}
```

In other words, the flag `resolved` will always be `true`. This in turn will make `if ( resolved || !copy.exists() )` equivalent to `if (true)`, thereby always executing the `FileUtils.copyFile(destination, copy)`. Hence, Maven will end up copying each and every snapshot dependency. For bigger projects, I can imagine this accumulates to significant IO, slowing down the build.

The attached patch proposes to compare the file's timestamp&size instead of using the `resolved` flag. This should also ensure that artifact-\<baseVersion>.jar matches the last resolved timestamp version but still avoids unnecessary file copies in case nothing changed.

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

A comment about the IT: I suggest to remove the checksum files (*.md5, *.sha1) from the test repository and just setup the repo to ignore the checksums. Files that are not vital for a test otherwise simply cause IO and slow down resource copying/extraction during testing. This is not relevant for a single test but we approach 400 ITs so there's a certain scaling factor to consider.

**Comment by jira-importer** [comment]
**[Maria Odea B. Ching](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oching)** commented

Ok, thanks for reviewing them Benjamin. I'll update the patches. Should I merge this to 2.2.x branch as well after I commit it to 2.1.x branch?

**Comment by jira-importer** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

> Should I merge this to 2.2.x branch as well
> Yes, please, I don't see a reason to not do so.

**Comment by jira-importer** [comment]
**[Maria Odea B. Ching](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oching)** commented

Fix committed in 2.1.x branch -r788583 and merged to 2.2.x branch -r788596.
Integration tests committed in core-integration-testing trunk -r788584.
